cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2016 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.cats.redis.cluster;
+
+import com.netflix.spinnaker.cats.agent.*;
+import com.netflix.spinnaker.cats.module.CatsModuleAware;
+import com.netflix.spinnaker.cats.redis.JedisSource;
+import com.netflix.spinnaker.cats.thread.NamedThreadFactory;
+import redis.clients.jedis.Jedis;
+
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/*
+ * The idea behind this scheduler is simple. Every agent it owns is always in one of two sorted sets,
+ * WORKING, or WAITING. If it is in the WORKING set, it's being run, if it's in the WAITING set, it is waiting to run.
+ * These sets are sorted by time, and the rank is time at which they are ready to run (in the WAITING set), or have
+ * timed out (in the WORKING set). Therefore, if we do a `zrangebyscore SETNAME -inf CURRENT_TIME`, we get a list
+ * of all agents that have expired. Those are then atomically pulled from the set, and run by the scheduler.
+ *
+ * The amortized cost of this scheduler is much lower than the original ClusteredAgentScheduler, since during each
+ * cache interval every key will only be removed from Redis once. If the interval is 60s, and the agent polls every 1s,
+ * we already have a 60 x 1 x # of clouddrivers factor of improvement.
+ */
+class ClusteredSortAgentScheduler extends CatsModuleAware implements AgentScheduler, Runnable {
+  private final JedisSource jedisSource;
+  private final NodeStatusProvider nodeStatusProvider;
+  private final ExecutorService agentWorkPool;
+  private final Integer parallelism;
+
+  private static Integer NOW = 0;
+  private static Integer INTERVAL = 30;
+  private static Integer TIMEOUT = 500;
+
+  private Map<String, AgentWorker> agents;
+  private volatile Integer runningAgents;
```

Comment:
Semaphore will block, no? That would break the behavior I have below in `saturatePool`.

And why `AtomicInteger` over `volatile`? I know very little Java